tests/Unit/DataStructures/Tensor/Expressions/Test_Contract.cpp

```diff
@@ -0,0 +1,540 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Framework/TestingFramework.hpp"
+
+#include <cstddef>
+#include <iterator>
+#include <numeric>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Expressions/Contract.hpp"
+#include "DataStructures/Tensor/Expressions/Evaluate.hpp"
+#include "DataStructures/Tensor/Expressions/TensorExpression.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Utilities/MakeWithValue.hpp"
+#include "Utilities/Requires.hpp"
+
+namespace {
+template <typename... Ts>
+void create_tensor(gsl::not_null<Tensor<double, Ts...>*> tensor) noexcept {
+  std::iota(tensor->begin(), tensor->end(), 0.0);
+}
+
+template <typename... Ts>
+void create_tensor(gsl::not_null<Tensor<DataVector, Ts...>*> tensor) noexcept {
+  double value = 0.0;
+  for (auto index_it = tensor->begin(); index_it != tensor->end(); index_it++) {
+    for (auto vector_it = index_it->begin(); vector_it != index_it->end();
+         vector_it++) {
+      *vector_it = value;
+      value += 1.0;
+    }
+  }
+}
+
+template <typename DataType>
+void test_contractions_rank2(const DataType& used_for_size) noexcept {
+  // Contract (upper, lower) tensor
+  // Use explicit type (vs auto) so the compiler checks the return type of
+  // `evaluate`
+  Tensor<DataType, Symmetry<2, 1>,
+         index_list<SpatialIndex<3, UpLo::Up, Frame::Inertial>,
+                    SpatialIndex<3, UpLo::Lo, Frame::Inertial>>>
+      Rul(used_for_size);
+  create_tensor(make_not_null(&Rul));
+
+  const Tensor<DataType> RIi_contracted =
+      TensorExpressions::evaluate(Rul(ti_I, ti_i));
```

Comment:
For my understanding, not a change request: what happens if you contract a (lower, lower) tensor? will the machinery work, or does it error?

Reply:
I ended up writing more than I expected, so if you're game, feel free to check out the details! If not, hopefully the **TLDR** right here answers enough.

**TLDR**:
Right now, you will get complaints if you don't have opposite valences in the `Tensor`'s index list, but you won't get complaints from solely doing `R(ti_i, ti_i)`, and this is because these two entities are disconnected right now and not checked for consistency in terms of valence or index type (spatial/spacetime), mostly because `TensorIndex` doesn't keep track of this information. My plan was for my next PR to be the one that incorporates valence and index type into `TensorIndex` so that things like `ti_i` and `ti_I` keep track of these things. Through making `TensorIndex` keep track of valence, specifically, `ti_i` and `ti_I` will then be distinguishable, and checks can be run.

**More details**
Right now, the generic index objects like `ti_I` or `ti_i` are disconnected from the `SpacetimeIndex`/`SpatialIndex` indices in the `Tensor`. So, right now, if I make a tensor expression like `Rul(ti_I, ti_i)`, there is nothing enforcing that the two indices in the `Rul` are indeed `SpatialIndex`s (since `ti_I` and `ti_i` are intended to represent spatial indices, as opposed to spacetime), and there is also nothing enforcing that the two indices' valences match the valences suggested by `ti_I` (implying upper valence) and `ti_i` (implying lower valence) - so, there is no checking that the first index in the `Rul`'s index list has upper valence and its index type is `SpatialIndex` (i.e. check for consistency with `ti_I`), and there is no checking that the second index has lower valence and its index type is `SpatialIndex` (i.e. check for consistency with `ti_i`). This is because right now, `TensorIndex`s only store a positive integer value, but don't store any information about valence or index type (yet!). And moreover, the upper and lower versions store the same value and are the same type, e.g. `ti_A.value` and `ti_a.value` are both `0`, and these two have the exact same type, `TensorIndex<0>`.  So, the important context here is: upper and lower versions of these generic `TensorIndex`s are indistinguishable at this point, and there aren't checks done to ensure that these representations are consistent with a `Tensor`s index list. The main reason for this is because valence and index type is not yet information kept track of by `TensorIndex`s, but this change is what I have on deck to submit as my next PR.

The only check that's done regarding whether or not indices provided can be contracted is in this assert [here](https://github.com/sxs-collaboration/spectre/blob/b9591b16b7445300026882c6e86edb5167d0262b/src/DataStructures/Tensor/Expressions/Contract.hpp#L99), which is defined [here](https://github.com/sxs-collaboration/spectre/blob/b9591b16b7445300026882c6e86edb5167d0262b/src/DataStructures/Tensor/Expressions/Contract.hpp#L22). You can see in the definition that the following checks are done for the `Tensor`s index list: do the `Tensor`'s indices have the same dimension, frame type, and index type (`SpacetimeIndex` or `SpatialIndex`) and opposite valence? So those checks are done, but there aren't any checks done to see if the generic `TensorIndex`s like `ti_I` and `ti_i` match the information in the `Tensor`'s index list.

**What this all means in terms of the original question...**
1. If the `indices_contractible` assert passes (dim, frame type, and index type of contracted indices are the same, and their valences are opposite), then nothing is going to complain if I do `Rul(ti_i, ti_i)` for the expression (and I could even do `Rul(ti_a, ti_a)` if I wanted, even though in the index list, we see `SpatialIndex` as the index type). So this compiles and  performs the contraction just fine:
```
Tensor<DataType, Symmetry<2, 1>,
         index_list<SpatialIndex<3, UpLo::Up, Frame::Inertial>,
                    SpatialIndex<3, UpLo::Lo, Frame::Inertial>>>
      R(used_for_size);
  create_tensor(make_not_null(&R));

  const Tensor<DataType> Raa_contracted =
      TensorExpressions::evaluate(R(ti_a, ti_a)); // valences and index type don't match above index_list
```

2. However, it _will_ complain if I do something wrong in the `index_list` for the `Tensor`, regardless of whether I have mixed casing (e.g. `R(ti_I, ti_i)`) or not (e.g. `R(ti_i, ti_i)`). The following doesn't compile because it fails the `indices_contractible` assert due to the two indices having the same valence:
```
Tensor<DataType, Symmetry<2, 1>,
         index_list<SpatialIndex<3, UpLo::Lo, Frame::Inertial>, // <-- switched this to Lo
                    SpatialIndex<3, UpLo::Lo, Frame::Inertial>>>
      R(used_for_size);
  create_tensor(make_not_null(&R));

  const Tensor<DataType> Raa_contracted =
      TensorExpressions::evaluate(R(ti_a, ti_a)); // valences and index type don't match above index_list
```

Reply:
This was very clear and helpful — thanks for taking the time to type it all out! 👍